### ISSUE TITLE
Enable query consolidation only by opt-in

### ISSUE THREAD

**Issue by aquarapid** [vitessio/vitess#7094]
Since Vitess query consolidation can break normal MySQL read committed guarantees outside of transactions, it can be difficult to strongly verify if an application can handle this before switching query consolidation on.  We can already switch on query consolidation on the replica, but not the master (via `-enable-consolidator-replicas` and `-enable-consolidator=false`).

However, some applications need to send queries to the master that can also benefit from query consolidation.  So one idea is to allow queries to opt into the consolidator by using a "Vitess comment", as we do for custom query timeouts, etc.  In this mode only queries that contain the comment will be eligible for consolidation.

**Comment by techwolf359** [comment]
+1, I can see this being beneficial for cases where you have slow changing data but high rates of reads. Typical caching works well until the cache fails, then you have a thundering herd of traffic to the database. Enabling consolidation on those specific queries would allow you one more layer of defense preventing the saturation on MySQL.

**Comment by dweitzman** [comment]
I think this is a dupe of https://github.com/vitessio/vitess/issues/4058

**Comment by Susurrus19** [comment]
+1. Going to raise this again as something that would be useful in cases where consistency is  paramount. With the flags being enabled in [10907](https://github.com/vitessio/vitess/pull/10907/files) we would like to implement this tech, but would like to do it selectively.

**Comment by harshit-gangal** [comment]
I think we should change the flag default to `--enable-consolidator=false --enable-consolidator-replicas`. This will make consolidator to be enabled only on replica and we can use the new approach to enable query based with #11080

**Comment by deepthi** [comment]
> I think we should change the flag default to `--enable-consolidator=false --enable-consolidator-replicas=true`. This will make consolidator to be enabled only on replica and we can use the new approach to enable query based with #11080 on primary

Changing the defaults will be a breaking change for anyone who is relying on the current behavior. I would not recommend it, especially since it is easy enough for those who want to opt-in to #11080 to change their configuration.

**Comment by dweitzman** [comment]
It's not breaking in the sense that it would cause issues with query results. If it causes issues for anyone, the issue would be increased tablet load.

I don't think the tablet version upgrade agreement promises that people would be able to do a non-incremental rollout across all tablets and expect identical performance to the prior release, so from my perspective at least this doesn't sound like a breaking change

**Comment by harshit-gangal** [comment]
The main concern here is consistency gurantees of a primary when running a query without a transaction is should always return the latest value. With running consolidator by default there can be instances where it can return stale result.

Here I would say correctness will take over a breaking change in the config.
Regarding performance regression or old behaviour (only or those who are using it already) can change back the setting.

**Comment by techwolf359** [comment]
I agree with @harshit-gangal here. We should never break accuracy for consolidation, which is why I suggested using query comments as a way to opt queries in. Today, consolidation is all-or-nothing.

**Comment by deepthi** [comment]
> The main concern here is consistency gurantees of a primary when running a query without a transaction is should always return the latest value. With running consolidator by default there can be instances where it can return stale result.
> 
> Here I would say correctness will take over a breaking change in the config. Regarding performance regression or old behaviour (only for those who are using it already) can change back the setting.

Ok, it makes sense to say that flags that affect correctness should be opt-in. We can make the change and document it in the release notes. We need to publicize it as widely as possible before it happens.

**Comment by deepthi** [comment]
cc @demmer @brirams

**Comment by brirams** [comment]
we've had consolidation on across primaries/replicas for a while and have been relying on its default. it'd be a nasty surprise to learn that it wasn't on one day. it's easy enough for us to set it explicitly(making that change now) but would consider this a "breaking" change even if the "breaking" here is a fix to a potential bug. i don't know how a deprecation cycle would work for changing the default value of a flag and that might be too long to continue running in this way but at a minimum, this should be communicated loudly and widely, as deepthi said.

**Comment by harshit-gangal** [comment]
For changing a flag default there is no deprecation cycle, the flag will continue to exist. For Slack I would suggest that you can start setting that flag explicitly from now onwards so that when you move to 15.0 there will be no surprise.
